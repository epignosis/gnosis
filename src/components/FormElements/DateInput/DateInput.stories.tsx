import React, { useState } from "react";
import { Story } from "@storybook/react";
import DateInputComponent, { DateInputProps } from "./DateInput";

export default {
  title: "components/Form Elements/Date Input",
  component: DateInputComponent,
  args: {
    id: "dateInput",
    label: "Date of birth",
    size: "md",
    inline: false,
    className: "dateInputStory",
  },
  argTypes: {
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
    value: {
      control: {
        type: "date",
      },
    },
  },
  decorators: [
    (Story: Story): JSX.Element => (
      <div style={{ maxWidth: 400 }}>
        <Story />
      </div>
    ),
  ],
};

export const DateInput: Story<DateInputProps> = (args) => {
  const [date, setDate] = useState(new Date());

  return <DateInputComponent {...args} value={date} onChange={setDate} />;
};
