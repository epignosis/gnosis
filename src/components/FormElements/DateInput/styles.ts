import { css, SerializedStyles, Theme } from "@emotion/react";
import { inputBaseStyles, inputContainerBaseStyles } from "../styles";
import { InputSize } from "@components/FormElements/Input/Input";

export const dateInput = (
  { formElements }: Theme,
  { size }: { size: InputSize },
): SerializedStyles => css`
  ${inputContainerBaseStyles({ block: true })};

  &.inline {
    width: auto;
  }

  input {
    ${inputBaseStyles(formElements, { block: true, size })};
  }

  .react-datepicker-wrapper {
    width: 100%;
  }
`;
