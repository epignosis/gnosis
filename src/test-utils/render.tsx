import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
