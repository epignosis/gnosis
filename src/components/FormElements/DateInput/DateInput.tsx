import React, { FC, ChangeEvent } from "react";
import { useResponsive } from "@umijs/hooks";
import DatePicker from "react-datepicker";
import { SerializedStyles } from "@emotion/react";
import { format, parseISO } from "date-fns";
import classNames from "classnames";
import { dateInput } from "./styles";
import "react-datepicker/dist/react-datepicker.css";
import { Input, Label } from "@components";
import { CalendarSVG } from "@icons/core";
import { InputSize } from "@components/FormElements/Input/Input";

export type DateInputProps = {
  value: Date;
  id?: string;
  size?: InputSize;
  label?: string;
  altLabel?: boolean;
  className?: string;
  inline?: boolean;
  onChange: (selectedDate: Date) => void;
};

const DateInput: FC<DateInputProps> = (props) => {
  const {
    value = new Date(),
    id,
    size = "md",
    className,
    label,
    altLabel = false,
    inline = false,
    onChange,
  } = props;
  const { sm } = useResponsive();
  const hasLabel = Boolean(label);
  const containerClassNames = classNames({
    "alt-label": hasLabel && altLabel,
    [className ?? ""]: true,
    inline,
  });
  const onChangeDate = (date: Date): void => {
    onChange(date as Date);
  };
  const onChangeMobileDate = (e: ChangeEvent<HTMLInputElement>): void => {
    const date = parseISO(e.target.value);
    onChange(date);
  };

  return (
    <div
      css={(theme): SerializedStyles => dateInput(theme, { size })}
      className={containerClassNames}
    >
      {hasLabel && <Label htmlFor={id}>{label}</Label>}
      {!sm ? (
        <input
          type="date"
          id={id}
          className={className}
          value={format(value, "yyyy-MM-dd")}
          onChange={onChangeMobileDate}
        />
      ) : (
        <DatePicker
          id={id}
          className={className}
          selected={value}
          onChange={onChangeDate}
          customInput={<Input size={size} iconAfter={CalendarSVG} />}
        />
      )}
    </div>
  );
};

export default DateInput;
