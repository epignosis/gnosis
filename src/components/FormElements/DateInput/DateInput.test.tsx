import React from "react";
import { format } from "date-fns";
import faker from "faker";
import DateInput from "./DateInput";
import { render, screen } from "@test-utils/render";

describe("<DateInput />", () => {
  it("renders correctly", () => {
    const mockFn = jest.fn();
    const labelTxt = faker.random.words();
    const date = faker.date.future();

    render(
      <DateInput
        id={faker.random.alphaNumeric()}
        label={labelTxt}
        value={date}
        onChange={mockFn}
      />,
    );

    const label = screen.getByText(labelTxt);
    const input = screen.getByLabelText(labelTxt);

    expect(label).toHaveTextContent(labelTxt);
    expect(input).toHaveValue(format(date, "MM/dd/yyyy"));
    expect(mockFn).not.toHaveBeenCalled();
  });

  it("matches snapshot", () => {
    const date = new Date("Thu Aug 19 2021 13:29:16 GMT+0300 (Eastern European Summer Time)");
    const { container } = render(
      <DateInput id="test-date-input" label="Some date" value={date} onChange={jest.fn()} />,
    );

    expect(container).toMatchSnapshot();
  });
});
