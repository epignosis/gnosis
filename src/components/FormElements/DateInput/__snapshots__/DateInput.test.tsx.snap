// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput /> matches snapshot 1`] = `
<div>
  <div
    class=" css-ow2xce-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-dateInput-dateInput"
  >
    <label
      class="css-xs6o3g-label-label"
      for="test-date-input"
    >
      Some date
    </label>
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <div
          class="valid with-suffix-icon css-76gtfa-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-inputContainer-inputContainer"
        >
          <div
            class="input-wrapper"
          >
            <input
              aria-describedby=""
              id="test-date-input"
              value="08/19/2021"
            />
            <span
              class="suffix-icon"
              data-testid="input-icon-after"
            >
              <svg-file-stub
                height="32"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
